### ISSUE TITLE
boost mp on Apple clang

### ISSUE THREAD

**Issue by sloriot** [CGAL/cgal#3816]
It seems there are some issues with `boost::multiprecision` (of at least boost 1.67) and Apple clang.

See for example here:
https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-4.14-Ic-160/Alpha_shapes_2/TestReport_cgaltester_x86-64_Darwin-18.2_Apple-clang-default_Release.gz

@mglisse Have you seen that already ?

**Comment by mglisse** [comment]
> @mglisse Have you seen that already ?

I don't think so. I'll have a look.

**Comment by mglisse** [comment]
On linux, with boost 1.67, I tried several versions of clang (6, 7, 8, 8-with-libc++) and couldn't reproduce. So I don't know how to debug that...

**Comment by mglisse** [comment]
https://github.com/boostorg/multiprecision/commit/43e414d60b669c52d5333e19d27814a62ed34463
Sounds like this should be fine when compiling in C++11 or C++14 mode, or when using boost 1.70+. Since it doesn't reproduce here, I won't look for a workaround (probably hard anyway). If you really want, all I can suggest is, for the 4.14 branch only (not master which uses C++14), to use Gmpq/Gmpz instead of mpq_rational/mpz_int with the affected compiler.

**Comment by lrineau** [comment]
There seems to be a regression for all MacOS users, with recent AppleClang versions. But it is not clear which combinations of AppleClang and Boost version are impacted.

Boost 1.70 will be released 10th April. Can we delay and see if any CGAL user complains?

**Comment by lrineau** [comment]
Most users will probably use the defaults of their platform. And for CGAL, `WITH_GMPXX` if `OFF` by default, and C++03 is the default of AppleClang 10.

**Comment by mglisse** [comment]
> So you suggest that I install GMPXX on the test platform?

No, we have to test mpq_rational somewhere...

> And for CGAL, `WITH_GMPXX` if `OFF` by default

Really, if I have gmpxx installed, by default cgal's cmake won't try to use it? Could you please change that?

**Comment by sloriot** [comment]
@lrineau What should we do with that issue?

**Comment by lrineau** [comment]
I do not know. @danston: Dmitry, does this bug still impact our testsuite platforms on MacOS?

**Comment by danston** [comment]
@lrineau Yes, when I am trying to compile examples and/or demos with Boost 1.70+, if I do not use `-DCMAKE_CXX_FLAGS='-DCGAL_DO_NOT_USE_BOOST_MP'` some examples fail, but I can check it in more detail. You can assign me to this bug. I will investigate.

**Comment by mglisse** [comment]
If you are still getting a problem with recent boost, then it means the issue I linked to is not the same, so you should isolate a small reproducer and report it to boost (if this particular boost package hasn't dropped C++03 compatibility yet).

**Comment by lrineau** [comment]
I agree with Marc. Dmitry, when you reproduce those errors, what that in the branch for 4.14, or in `master`?

**Comment by danston** [comment]
Actually, it was in 5.01 if I am not mistaken.

**Comment by lrineau** [comment]
And thus it was with C++14.

**Comment by danston** [comment]
Ok, I have investigated the problem.

Indeed, the problem with the boost mp that I encountered is not the same as the one of this issue, however it is related. My problem causes some packages to fail, when compiling, with errors of the type:

```C++
boost/multiprecision/gmp.hpp: error: no member named "_mp_size" in "__mpf_struct"
boost/multiprecision/gmp.hpp: error: no template named "_mp_size";
```

The two above come from the `Poisson_surface_reconstruction_3` package. All CGAL affected examples are:
```
	517 - compilation_of__poisson_reconstruction (Failed)
	863 - compilation_of__simple_mcfskel_example (Failed)
	865 - compilation_of__simple_mcfskel_sm_example (Failed)
	867 - compilation_of__MCF_Skeleton_example (Failed)
	868 - compilation_of__MCF_Skeleton_sm_example (Failed)
	870 - compilation_of__segmentation_example (Failed)
```
All above fails in CGAL-5.0, CGAL-5.0.1, CGAL-5.0.2 and master with boost 1.70+.

The problem has already been discussed [here](https://stackoverflow.com/questions/57857572/why-does-boost-log-break-boost-gmp-multiprecision) and it is not a CGAL or boost problem but rather a [Clang Parser Bug](https://bugs.llvm.org/show_bug.cgi?id=43266). It was also reported in boost [here](https://github.com/boostorg/log/issues/90).

I tried to compile all above with `Apple clang version 11.0.0 (clang-1100.0.33.17)` and it failed. However, the new version `Apple clang version 11.0.3 (clang-1103.0.32.59)` includes the fix for that issue and everything works just fine!

When using an older clang version, the fix that I applied to compile all examples was adding 
```cmake
-DCMAKE_CXX_FLAGS='-DCGAL_DO_NOT_USE_BOOST_MP'
```
suggested by @lrineau.

**Comment by lrineau** [comment]
Since the bug is fixed in Apple Clang, shall we close this issue?

**Comment by danston** [comment]
@lrineau I think so.

**Comment by sloriot** [comment]
Should there be a workaround to disable boost mp on such platforms?

**Comment by danston** [comment]
@sloriot There is one, I mentioned above. Or you propose to detect the wrong version of Apple Clang and then disable boost mp for it?

**Comment by sloriot** [comment]
I mean something automatic :)

**Comment by danston** [comment]
@sloriot Ok, I will have a look at it and see if it is possible.

**Comment by danston** [comment]
I have proposed a fix for this issue in the PR #4710. Please have a look if this is a valid fix. Thanks!
